chanzuckerberg/idseq-web
# Cache report info in redis

# Description

This is a replacement for https://github.com/chanzuckerberg/idseq-web/pull/2219 using Redis instead of file cache and low level cache instead of action level cache. 

It fixes the bad perf of the sample report page by caching the http response. The browser HTTP client-side caching behavior is also optimized.

My tests of a large sample show 
1) report_info time decreased from 8s to 1s by server side caching
2) report_info time decreased to less than 1s by client side caching

I refactored some of the `report_info` and child methods to make the param dependencies more clear. I refactored the `ReportHelper` class to be mostly private class methods because that's what they were--they didn't need instance state or external access.

I chose to precache all `background_id` after all because Redis should act as a LRU cache when `expires_in` is enabled, so unused data will be evicted. 

We currently have 6 GB free in a cache.m3.large according to https://us-west-2.console.aws.amazon.com/elasticache/home?region=us-west-2. 

We I chose to share the Redis machines with Resque because Redis will act both as LRU and persistent in the default `volatile-lru` mode. 

I chose the default expires to be one month, on the assumption that anything worth seeing will be viewed within a month of precaching. 

I expanded the definition of `report_ts` to be the max `updated_at` of all child objects of a pipeline run, such as `taxon_counts`. This should capture all changes in underlying data, now and going forward. 

The git version is no longer in the cache key though I left it in the params for debugging. 

I removed the `sort_by` and `scoring_model` as action params, because they were always the defaults anyway. 

# Test and Reproduce

Automated tests
`rails t`

Test caching
1. Open a report such as http://localhost:3000/samples/12303
2. See SQL queries in rails log 
3. Reload the report 
4. See cache line in rails log such as 

```
ActiveSupport::Cache::FileStore : Cache read: /tmp/rails_cache_store/BD7/1E3/views%2Flocalhost%3A3000%2Fsamples%2F10399%2Freport_info.json%3Fbackground_id%3D29%26git_version%3D%26pipeline_version%3D3.3%26report_ts%3D1548725583%26sort_by%3Dnt_aggregatescore ({:expires_in=>365 days})
```

5. Verify that the decoded URL is correctly unique in param values
```/tmp/rails_cache_store/BD7/1E3/views/localhost:3000/samples/10399/report_info.json?background_id=29&git_version=&pipeline_version=3.3&report_ts=1548725583&sort_by=nt_aggregatescore```

6. Verify cache on further reloads of report page

7. Verify cache miss on changing background


Test precaching:

1.  In `rails console`, load a pipeline run
2. Precache  method and see the following:

```
irb(main):002:0> PipelineRun.find(18660).precache_report_info!
...
[2019-05-07T19:28:38] DEBUG Rails : Precaching URL http://localhost:3000/samples/12295/report_info?background_id=93&format=json&git_version=&pipeline_version=3.3&report_ts=1549325153&scoring_model=aggregate_score&sort_by=nt_aggregatescore
=> #<Tempfile:/tmp/open-uri20190507-144-u0xcg3>
```
3. Run it again and see a cache hit in the rails logs


Overall perf testing:

I'll review the stats before and after launch to prod in the airbrake dashboard.
